Span collector metrics

pbetkier opened the pull request:
Allows to monitor span collector's performance by implementing a single interface. Leaves the developer with free choice of which parameters to monitor, how to call these metrics and which metrics library to use.

What do you think?

kristofa commented:
@pbetkier Thanks for this! This is something that is really useful. Especially measuring when the back-end is not keeping up and we are dropping spans. Unfortunately I wil only be able to look at it in more detail beginning of next week.

kristofa commented:
@pbetkier I would make the implementation more generic. Instead of having an interface which is tightly coupled to the current Scribe implementation I would make more generic interfaces in a separate sub project: `Counter`, `Gauge`, 'Histogram` and probably a `MetricsFactory` to create the 3 metric types.

Reason is that we expect to have multiple span collector implementations soon (Kafka, http) but those might not follow same design as the current Scribe one. They might for example not batch spans prior to sending, not us an internal queue,... so they could have different metrics to report.

I'm willing to spend some time later this week to come up with an initial implementation. Do you think this makes sense?

pbetkier commented:
@kristofa What I like about the mechanism that I suggested is the flexibility for the developer: I can decide which parameters to add to my metrics registry and which to ignore, how to call them and perhaps implement a different strategy of notification for some of them, e.g. logging. That would require creating tailored interfaces for each span collector implementation though and a bit of work for the developer to start monitoring things.

I also see the power of an abstraction over metrics that you propose – it would be very easy to setup for the developer and allow him to switch between different span collectors easily. We could even create ready implementations for the most commonly used metrics libraries, like dropwizard metrics.

So it's a trade-off. I like to have control over how things are monitored so I prefer the original solution :) But your idea also makes sense, so go ahead with implementing it you value it more.

adriancole commented:
My 2p here is that we should only specify the minimum possible. For example the input to the histogram vs the histogram api. I'm even troubled by the queue size guage. This only matters when something else occurs in relation to the queue size. IOTW, we could log the size when that happens (or even allow folks to provide a queue)

In some private implementations of  ML stuff, I have seen this as simple as having classes that perform work leave unimplemented methods for things like incrementing. Ex a simple subtype could bind someone's favorite telemetry instrumentation vs making our own wrappers.

Brave is simple enough that basically I don't see value in cluttering it with an abstraction layer of any sort, solely to provide meters. If we did, would really be best to do the least.

kristofa commented:
@adriancole Can you clarify a bit more? I'm not sure I understand what you want to say :)

I was thinking about something like this:

```
interface Telemetry {
 Counter counter(String name)
 Gauge gauge(String name)
 Histogram histogram(String name)
}

interface Counter {
  void inc()
  void inc(double value)
}

interface Gauge {
  void set(double value)
}

interface Histogram {
 void observe(double value)
}
```
We would pass a `Telemetry` instance in the `ScribeCollector` and it can create the Counter / Gauge / Histogram instances that are needed.

adriancole commented:
just wondering how many things you are actually measuring :) Ex. I had guessed there would be less than 10 core metrics in brave. Does this require a type system?

Reason I say that is if you don't then you just expose the actual events ex `incrementSpansReceived(int)` which doesn't require a binding to the namespace scheme an interface hierarchy like above implies.

adriancole commented:
guess what I'm wondering if we can just expose what we need to record vs creating a type system around the abstract act of recording.

adriancole commented:
Here's a good reason. Notice you chose double for all the values. You only need to do such a thing when creating abstract systems. spans are always maximum an int (as they are in a collection who's max size is int). histogram with a single idea of double requires a lot of explanation, especially how to merge this thing? We can avoid all of these things by just concretely exposing what's going on as method signatures.

If brave were a complex system, maybe we should have our own telemetry api, just I don't understand how it helps tell the story of what people will be interested in wrt brave, and certainly it limits implementation.

adriancole commented:
very long winded way of saying that I'm more in favor of the direction of this pull request (notable exception is SpanCollectorGauge type), vs creating a telemetry api only brave uses. :)

kristofa commented:
@adriancole Yep, understand :) My proposal might be a case of YAGNI :)
Fair enough. I don't have very strong arguments to put against that so I'm fine with that.

Instead of the `Gauge` we could have a method like: `setCurrentQueueSize(int)`?

adriancole commented:
yeah basically I'd love whatever interface we have to tell the story
of notable events (you should bind to telemetry and probably alarms)!

Ex. if we do it like this, we can put javadoc literally
`onCurrentQueueSize(int)` which says watch out for buildup!

by using methods, we can also link to correlate things we may want to
advise people to alarm on.

Can't do that with abstractions, right? The beauty is that brave is
"experienced enough" to have the advantage of concrete guidance. let's
use it!

adriancole commented:
ps `onCurrentQueueSize(int)` vs `incrementQueueSize(int)` is a discussion to have :) personally, I like the increment/decrement approach. as long as it is consistent, you can implement a snapshot of "current" with that.

adriancole commented:
don't mean to drag queue size snapshot into the weeds, as I know it is pros vs cons. just noting things we can toss around.

kristofa commented:
@pbetkier what do you think about the remarks by @adriancole ?
Can you do these changes?

pbetkier commented:
I agree with @adriancole, very insightful remarks. He convinced me that we
should replace a gauge with a onCurrentQueueSize method. I'm not sure about
increment / decrement naming though, maybe I'm just not used to that :)

So what do we change in the PR apart from replacing the gauge? Do you agree
with the events that I exposed to monitor, or would you add or remove
something?

I'll be happy to update the PR, but it'll happen after 11 Nov as I don't
have computer access right now.

On Friday, October 30, 2015, Kristof Adriaenssens <notifications@github.com
<javascript:_e(%7B%7D,'cvml','notifications@github.com');>> wrote:

> @pbetkier <https://github.com/pbetkier> what do you think about the
> remarks by @adriancole <https://github.com/adriancole> ?
> Can you do these changes?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openzipkin/brave/pull/101#issuecomment-152453462>.
>

kristofa commented:
@pbetkier Start with the ones that you really want to inspect right now. I think the events you defined make sense. It is fine to keep this pr open till you are available to finish it up. Thanks @pbetkier  and @adriancole!

pbetkier commented:
I've updated my PR by removing the gauge and simplifying the exposed notifications. Please have a look.

I'm still thinking of allowing developers to provide their own queue instead of publishing `onQueueSizeChanged` notifications though. That would simplify the span collector code and make the API more useful in my opinion. I will propose such change in a separate commit so that we have a base for discussion.

adriancole commented on brave-spancollector-scribe/src/main/java/com/github/kristofa/brave/scribe/EmptyScribeSpanCollectorNotificationsHandler.java:
```diff
@@ -0,0 +1,24 @@
+class EmptyScribeSpanCollectorNotificationsHandler implements ScribeSpanCollectorNotificationsHandler {
+
+    @Override
+    public void onQueueSizeChanged(int currentQueueSize) {
```
I'd prefer this to be more explicit. Ex instead of the somewhat vague of queue-size changed, make this an event:
onSpanAccepted, onSpanProcessed?

adriancole commented on brave-spancollector-scribe/src/main/java/com/github/kristofa/brave/scribe/ScribeSpanCollectorNotificationsHandler.java:
```diff
@@ -0,0 +1,35 @@
+     * @param batchSize the size of the spans batch that was sent.
+     * @param failed true, if sending has resulted in a failure.
+     */
+    void onSpansSent(long durationInMillis, int batchSize, boolean failed);
```
for callbacks that are intended for metrics, probably good to think about an example of how these would be exported. Ex. this particular callback seems cluttered with mixed signals. This would be easy to see when we try to explain how to map this to a timeseries metric.

adriancole commented on brave-spancollector-scribe/src/test/java/com/github/kristofa/brave/scribe/CountingScribeSpanCollectorNotificationsHandler.java:
```diff
@@ -0,0 +1,53 @@
+    public void onSpansSent(long durationInMillis, int batchSize, boolean failed) {
+        spansSentEvents.incrementAndGet();
+        if (failed) {
+            sendingFailures.incrementAndGet();
```
ex. this sends signals of mixed granularity. for example spansRejected is per span and these are ... who knows? I'd prefer signals to be of the same grain, and use atomic addition if we want to increment in bulk

adriancole commented:
Thanks for the next round. made some comments about a preference towards
same grain as the anchoring concern is spans not "send events". ex a
multithreaded app may arbitrarily have multiple send events, but the
collection tier only cares about spans. We'd probably be best with choosing
a grain that is consistent across the tracing service.

pbetkier commented:
@adriancole I like your remarks about granularity, this indeed is something worth keeping consistent, especially that such API would be more universal when thinking about other spancollector implementations. So I see the value in introducing `onSpanAccepted`, `onSpanProcessed`. And also `onSpanFailedToProcess` – to report a span dropped due to connection failure.

But I don't agree that we should only care about spans. Queue size and logging latency are technical metrics that give clear information about operational issues and their source. Also helpful for tuning the parameters. While the developer could deduce the current queue size by doing the maths on other notifications, that's just unnecessary complexity for him in my opinion.

So I would add the 3 notifications that I mentioned above and keep `onQueueSizeChanged` (or alternatively make it possible to provide own queue). I would also replace the `onSpansSent` method with a method notifying about logging durations. What do you think?

adriancole commented:
Let's bear in mind that coding to this api will be a seldom occurrence, so
developer burden for composing things, when they care to, is going to be
limited to a subset of an already small number of developers. It may only
be 3 people!

Again, I think that for each scenario we dream up, we *must* back it with
actual metrics, not hypothetical ones. So, if you really think duration is
going to be one, please list what those are and how you would alert on it.

We owe it to ourselves to make the simplest api possible (ie the least
hooks possible), as often code submitted is written once and never
maintained after.

adriancole commented:
Basically long story short is that I think we still haven't described well enough the feature.

I don't think flexible ways of adding queues, or conveniently adding arguments in the same method call is a feature.

Knowing when buildup happens, or that our key metric of spans/second dropped is a feature. Doing that without causing a performance problem is hard as we've not created benchmarks, yet, but that's still step 2.

Here's an example of a fairly clean namespace even if there's at least 3 grains, Api calls (PutRecord/s), Records and Bytes: http://docs.aws.amazon.com/kinesis/latest/dev/monitoring-with-cloudwatch.html

Want to give a try at documenting how to apply things in this hook to alerts or action?

adriancole commented on brave-spancollector-scribe/src/main/java/com/github/kristofa/brave/scribe/ScribeSpanCollectorNotificationsHandler.java:
```diff
@@ -0,0 +1,35 @@
+     *
+     * @param rejectedSpan the rejected span.
+     */
+    void onSpanRejected(Span rejectedSpan);
```
ps probably don't need the Span parameter to increment a counter. ex `incrementDroppedSpans()` or if we want to call out only dropped spans due to queue rejection (vs marshaling, timeouts or other reasons), `incrementRejectedSpans()`

adriancole commented on brave-spancollector-scribe/src/main/java/com/github/kristofa/brave/scribe/EmptyScribeSpanCollectorNotificationsHandler.java:
```diff
@@ -0,0 +1,24 @@
+/**
+ * Empty implementation ignoring all notifications.
+ */
+class EmptyScribeSpanCollectorNotificationsHandler implements ScribeSpanCollectorNotificationsHandler {
```
nit: prefer Noop or Null prefix

pbetkier replied:
I prefer Noop as well, but wanted to be consistent with `EmptySpanCollector`.

adriancole commented on brave-spancollector-scribe/src/main/java/com/github/kristofa/brave/scribe/SpanProcessingThread.java:
```diff
@@ -83,8 +86,8 @@ public Integer call() {
                     subsequentEmptyBatches++;
-
                 } else {
+                    notificationsHandler.onQueueSizeChanged(queue.size());
```
if we ran a benchmark, this would pop up depending on the type of queue in use. This is why I like increment instead.

pbetkier replied:
touché :)

adriancole commented on brave-spancollector-scribe/src/main/java/com/github/kristofa/brave/scribe/SpanProcessingThread.java:
```diff
@@ -105,9 +108,10 @@ public Integer call() {
     private void log(final List<LogEntry> logEntries) {
         final long start = System.currentTimeMillis();
         final boolean success = log(clientProvider.getClient(), logEntries);
```
if we were going to use System.currentTimeMillis() to calculate this (which would be more expensive and less precise than System.nanoTime()), we'd probably want to do so conditionally, as if there's a Noop notifier, we'd be creating overhead for no reason.

adriancole commented on brave-spancollector-scribe/src/test/java/com/github/kristofa/brave/scribe/CountingScribeSpanCollectorNotificationsHandler.java:
```diff
@@ -0,0 +1,53 @@
+
+import java.util.concurrent.atomic.AtomicInteger;
+
+class CountingScribeSpanCollectorNotificationsHandler implements ScribeSpanCollectorNotificationsHandler {
```
something like this is probably better in the test package, as it is an interesting example, but not actually used, and easy to recreate. I'd expect monitoring bindings to be direct, ex to a codahale or otherwise library, as that allows the most relevant features to be used.

pbetkier replied:
I don't understand, this class is in the test sources, used in the tests to verify that the notifications are indeed published. It's doesn't serve as an example on how to bind it to a metrics library in production code.

adriancole replied:
ack. somehow I misread this as in the main tree

adriancole commented on brave-spancollector-scribe/src/test/java/com/github/kristofa/brave/scribe/CountingScribeSpanCollectorNotificationsHandler.java:
```diff
@@ -0,0 +1,53 @@
+    private final AtomicInteger spansRejected = new AtomicInteger();
+    private final AtomicInteger spansSentEvents = new AtomicInteger();
+    private final AtomicInteger sendingFailures = new AtomicInteger();
+    private volatile int lastQueueSize;
```
We should make a benchmark project, so stuff like this is of more obvious impact. Wherever we try to track state, we'll likely inherit a performance hit. In this case, it is flushing to main memory on updates to these fields.

pbetkier replied:
You're right, but performance was not my concern here, it's just for checking the notifications logic in the unit tests. This class has no other intention.

adriancole commented on brave-spancollector-scribe/src/test/java/com/github/kristofa/brave/scribe/ScribeSpanCollectorTest.java:
```diff
@@ -93,7 +82,39 @@ public void testCollectWithDefaultAnnotation() throws TTransportException {
+    }
 
+    @Test
+    public void testNotifyOfSpanCollectionEvents() throws TTransportException {
```
good start on a feature test. This is where we can explain what the callbacks mean

adriancole commented on brave-spancollector-scribe/src/test/java/com/github/kristofa/brave/scribe/ScribeSpanCollectorTest.java:
```diff
@@ -93,7 +82,39 @@ public void testCollectWithDefaultAnnotation() throws TTransportException {
+        assertEquals(0, eventsHandler.getTotalSpansRejected());
+    }
+
+    private void collectSingleSpan(ScribeSpanCollector zipkinSpanCollector) {
```
I'd probably remove this method and inline it in the test, particularly as you can use try-with-resources on ScribeSpanCollector

pbetkier replied:
agreed

pbetkier commented:
I think the difference in our approaches lies in defining what this feature we're working on is. My original idea was to expose hooks a developer can use to implement his own decisions on what and how to monitor. It'd be up to him if he wants to be alerted by X or Y. Or perhaps he prefers to log stuff instead of using some metrics library. My thinking was that the audience of this library is limited to a few toolmakers and they, seeing how span collector works, are able to make such decisions for themselves. That's why I didn't want to think in terms of incrementing some counters and wanted to provide richer notifications.

But perhaps you're right that we should go further and be opinionated on what and how to monitor, solving this puzzle once and for all. Let's move our focus from the code for a second and try to think what possible failure scenarios do we have and what metrics and alerts do we need for solving them. Once we converge on this, making code changes will be much easier for me.

My view on this:
```
- failure: rejecting incoming spans due to queue being full
    - alert: rejected spans meter > 0 (or some more forgiving threshold)
    - possible reasons:
         - reason: a peak in incoming spans
           observed to learn why: a peak in spans accepted meter
         - reason: increased latency of sending the spans
           observed to learn why: a peak in latency of sending timer

- failure: discarding spans that couldn't be sent due to connection problem
    - alert: discarded spans meter > 0 (or some more forgiving threshold)
    - possible reasons:
         - reason: connection timeout
           observed to learn why: failed requests meter > 0 and plateau near socket timeout value on higher percentiles of sending timer
         - reason: connection broken
           observed to learn why: failed requests meter > 0 and no plateau near socket timeout value visible, probably latency lower than usual on median of sending timer
```

So what I need is a spans rejected meter, spans accepted meter, spans discarded meter, scribe delivery timer and scribe delivery failure meter. I still think it would be nice to view how we're lagging with processing work, even if we're not yet rejecting the spans (e.g. for tuning the configuration parameters): but if not monitoring the queue size directly, then we need spans accepted and spans processed (either delivered or discarded) counters to subtract.

What is your take on this? How would your similar list look like?

adriancole commented:
I'm concerned with this as it reads like guessing. Can you clarify if these
signals (and this work) is related to production outages where other
monitoring tools failed because the root cause was obfuscated in Brave?

@kristofa @eirslett you've certainly had experience. What do you
continually burn your fingers on, which would lead you to want a
brave-specific telemetry api? What are your top-3 failures that eat your
time today, which would eat less if we had brave-specific
meta-instrumentation?

Reason I ask is that unlike some libraries, Brave is 2.5 years old. There
*must* be existing prod support concerns we can leverage vs guessing.

adriancole commented:
ps Piotr, I bet it isn't obvious through my comments that I'm
interested in this work and your effort towards it. I am happy about
both.. I mainly want your effort behind a feature that is used by
folks besides you, as otherwise, if could just live in a fork.

Feel free to hop on https://gitter.im/openzipkin/zipkin if you want
real-time conversation, noting Eirik and Kristof are in central europe
time.

pbetkier commented:
I appreciate your remarks :) Indeed it would be great to hear some experiences from the battle-front. We're just introducing tracing on production, so no stories from my side.

kristofa commented:
@pbetkier @adriancole I wrote a message about our production experiences with Zipkin and what I think we should monitor here: https://gitter.im/openzipkin/zipkin

pbetkier commented:
@eirslett just a kind reminder, do you have some failure stories to share that would help us shape the metrics API for span collectors? Do you even see the need for monitoring the span collector's behavior?

pbetkier commented:
Quote from kristofa on gitter: 
> By now I have more experience with using Zipkin (through Finagle) in a production environment as with using Brave but the main issues we encountered so far are:
Cassandra filling up because it is not keeping up with removing expired data. This became obvious by monitoring Cassandra and got fixed by lowering sample rate (=less data) and increasing compaction throughput.
Dropped spans observed by zipkin-collector counter collector.pushback and Finagle counter try_later. This is quite important to look into because if spans are dropped traces do not represent reality which can be worse then having no tracing at all. I guess this can be fixed by having more zipkin-collector instances, tuning/increasing size of Cassandra cluster or lowering sample rate. We increased Cassandra cluster once and also lowered sample rate.
So I would say the most important metrics we want from Brave is counters with successfully delivered spans and dropped spans.

I suggest we go with measuring rejected, delivered and dropped spans then. Notifications API would expose 3 methods called like `incrementSpansDelivered()`. We suggest the developer registered corresponding meters in his registry and mark them when these increments occur.

What's your take on that @adriancole?

adriancole commented:
Thanks for keeping this alive!

My 2p

I think dropped vs accepted makes sense. I'm still apprehensive about a
separate counter for dropped (for any reason) vs rejected (to queuing
choice). I'd prefer a single dropped counter for now.

There's some nuance to what we indicate as delivered/accepted. Ex we can
count "accepted" as absolute number sampled. Or we can count "delivered" as
absolute number - dropped

I think the term "delivered" is a bit of a lie as we've no idea they were
in fact delivered. I like "accepted" better as it is something the user can
control. Ex sample less means accept less.

pbetkier commented:
Right, we could drop rejected counter and have one counter for spans dropped for any reason as the details on what happened are available in the logs anyway.

Do I understand your idea with accepted counter correctly – if a span gets rejected by the queue, it is marked both as accepted and dropped?

adriancole commented:
Right, my take is we increment accepted before any transport attempt is
made, and dropped if any exception occurs.

So 100 accept and 40 drops means 40% known loss.

@kristofa wdyt?

kristofa commented:
@pbetkier @adriancole That sounds good!

pbetkier commented:
I've updated my PR implementing the ideas we discussed, please have a look @kristofa @adriancole.

adriancole commented on brave-spancollector-scribe/README.md:
```diff
@@ -15,6 +17,17 @@ This approach has again been chosen to minimize the impact on the application. H
+counters in your metrics tool. When a span gets dropped, the reason is written to the application logs.
+The number of spans sent to the target collector can be calculated by subtracting the dropped count from the accepted count.
+
+Refer to `DropwizardMetricsScribeCollectorMetricsHandlerExample` for an example of how to integrate with 
```
Can link to the actual file

adriancole commented on brave-spancollector-scribe/src/test/java/com/github/kristofa/brave/scribe/ScribeReceiver.java:
```diff
@@ -81,4 +77,11 @@ public ResultCode Log(final List<LogEntry> messages) throws TException {
         return result;
     }
 
+    public void setDelayMs(int delayMs) {
```
Id revert as this is a little more ceremony than I'd expect in a test class

pbetkier replied:
When you say revert you mean replace it with what?

adriancole replied:
Revert I mean just go back to accessing the field (similar to how the code
was originally). Since it is a test class, this sort of defensive
programming seems distracting, IMHO.

pbetkier replied:
I see, I thought you meant removing the setter altogether.

adriancole replied:
I did :) but anyway I don't know how much attention this detail is worth.
Im fine either way

pbetkier replied:
The field is private, originally it was provided in the constructor (it was dead code though, nobody used this feature). I'd prefer to set it via setter where needed to have a common scribe server setup in my newly added test class.

adriancole replied:
I don't feel too strongly, just that setters are boilerplatey. If I were
writing this, I'd do it differently, but at this point we've given this
particular code far too much luxury of attention for its value. Let's move
on.

kristofa commented:
@pbetkier @adriancole thanks! I think we can merge this.

When we have other collectors (eg Kafka) and `ScribeCollectorMetricsHandler` has been put to use and proven to be valuable we can move the interface to `brave-core` and rename it to for example `CollectorMetricsHandler`. At that point we can use it in all collector implementations. 

But for now this is a good start.

kristofa commented:
@pbetkier @adriancole brave 3.2.0, which includes this pr, is released and should appear in Maven Central soon. Release notes will follow later, probably tomorrow.
